### PR TITLE
CDAP-7336 Return 400 instead of 500 if the route config is not valid

### PR DESCRIPTION
**Pull request by gokulavasan**
JIRA : https://issues.cask.co/browse/CDAP-7336

Build : http://builds.cask.co/browse/CDAP-RUT183-1

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/RouteConfigHttpHandler.java [hunk 1]
```diff
@@ -78,7 +78,14 @@ public void storeRouteConfig(HttpRequest request, HttpResponder responder,
                                @PathParam("service-id") String serviceId) throws Exception {
     ProgramId programId = Ids.namespace(namespaceId).app(appId).service(serviceId);
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/RouteConfigHttpHandler.java:80]: unrelated, but can you change this to just use `new ProgramId`?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/RouteConfigHttpHandler.java [hunk 2]
```diff
@@ -78,7 +78,14 @@ public void storeRouteConfig(HttpRequest request, HttpResponder responder,
...
     Map<String, Integer> routes = parseBody(request, ROUTE_CONFIG_TYPE);
-    routeStore.store(programId, new RouteConfig(routes));
+    RouteConfig routeConfig;
+    try {
+      routeConfig = new RouteConfig(routes);
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/RouteConfigHttpHandler.java:84]: would this be simplified if the RouteConfig constructor threw a BadRequestException?

**Review by bdmogal — COMMENTED** [review]
Couple comments. Rest LGTM

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/RouteConfigHttpHandler.java (hunk 2 above)
Comment (gokulavasan) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/RouteConfigHttpHandler.java:84]: BadRequestException seems to be tied to HttpErrorStatusProvider. RouteConfig doesn't need to know about HttpErrorStatusProvider?

Comment (bdmogal) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/RouteConfigHttpHandler.java:84]: what do you mean by tied? I guess HttpErrorStatusProvider is just another capability provided by BadRequestException, right?

**Comment by ecfm** [comment]
lgtm

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/RouteConfigHttpHandler.java (hunk 2 above)
Comment (gokulavasan) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/RouteConfigHttpHandler.java:84]: BadRequestException is related to our Http handler logic. Why should RouteConfig constructor throw that exception?

Comment (bdmogal) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/RouteConfigHttpHandler.java:84]: Is it though? It seems like a generic exception that anyone could throw - providing HTTP response status is just additional capability? Service classes (that HTTP handlers delegate to) use these exceptions - technically these classes could be used anywhere. I thought there were jiras to even have the `Store` interface throw these exceptions.

Don't feel too strongly about this though - just a nit - code could be cleaner that way.

**Pull request merged**